src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,184 @@
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
+        }
+
+        protected override void PostPersist(ApplicationEngine engine)
+        {
+            base.PostPersist(engine);
+            (UInt160 Account, BigInteger GAS)[] nodes = null;
+            foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            {
+                //Filter the response transactions
+                OracleResponse response = tx.GetAttribute<OracleResponse>();
+                if (response is null) continue;
+
+                //Remove the request from storage
+                StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+                OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+                engine.Snapshot.Storages.Delete(key);
+
+                //Remove the id from IdList
+                key = CreateStorageKey(Prefix_IdList).Add(GetUrlHash(request.Url));
+                IdList list = engine.Snapshot.Storages.GetAndChange(key).GetInteroperable<IdList>();
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                if (nodes == null) nodes = GetOracleNodes(engine.Snapshot).Select(p => (Contract.CreateSignatureRedeemScript(p).ToScriptHash(), BigInteger.Zero)).ToArray();
+                if (nodes.Length > 0)
+                {
+                    int index = (int)(response.Id % (ulong)nodes.Length);
+                    nodes[index].GAS += OracleRequestPrice;
+                }
+            }
+            if (nodes != null)
+            {
+                foreach (var (account, gas) in nodes)
+                {
+                    if (gas.Sign > 0) GAS.Mint(engine, account, gas);
+                }
+            }
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData, long gasForResponse)
+        {
+            //Check the arguments
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength
+                || gasForResponse < 0_10000000)
+                throw new ArgumentException();
+
+            //Mint gas for the response
+            engine.AddGas(gasForResponse);
+            GAS.Mint(engine, Hash, gasForResponse);
+
+            //Increase the request id
+            StorageItem item_id = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_RequestId));
+            ulong id = BitConverter.ToUInt64(item_id.Value) + 1;
```

Comment:
The request Id it's predictable and it's used for reward a specific Oracle Node, if the owner of a exchange (or someone who produce a lot of tx) own a Oracle node can send transactions only if this node will be rewarded.

Reply:
This will not cause uneven distribution.

Reply:
If we make a hash, it could be uneven, but statistically should be fair.